File: apple/Cabalmail/ViewModels/FeedHealth.swift
import Foundation
import CabalmailKit

/// The fetcher's view of a feed, reduced to what a sidebar badge can say
/// (RSS plan, phase 10, pulled forward 2026-09-10 after an OPML import
/// left two silently empty feeds). Thresholds match the plan: three
/// consecutive failures is worth a warning; twenty, or a dead letter, means
/// the fetcher has stopped trying.
enum FeedHealthLevel: Equatable {
    case healthy
    case failing(Int)
    case stopped

    var symbol: String? {
        switch self {
        case .healthy: return nil
        case .failing: return "exclamationmark.triangle.fill"
        case .stopped: return "xmark.octagon.fill"
        }
    }

    /// Short wording for the badge's accessibility label and tooltip.
    var summary: String? {
        switch self {
        case .healthy: return nil
        case .failing(let count): return "Failing: \(count) fetches in a row"
        case .stopped: return "Stopped: the fetcher gave up on this feed"
        }
    }
}

enum FeedHealth {
    static let warningThreshold = 3
    static let stoppedThreshold = 20

    static func level(for feed: RssFeedSummary?) -> FeedHealthLevel {
        guard let feed else { return .healthy }
        if feed.deadLettered || feed.consecutiveFailureCount >= stoppedThreshold { return .stopped }
        if feed.consecutiveFailureCount >= warningThreshold { return .failing(feed.consecutiveFailureCount) }
        return .healthy
    }

    /// One line for the item list's header while a feed is unwell: the
    /// level, then the fetcher's own words when it left any.
    static func headline(for feed: RssFeedSummary?) -> String? {
        guard let feed, let summary = level(for: feed).summary else { return nil }
        let detail = feed.lastError.trimmingCharacters(in: .whitespacesAndNewlines)
        return detail.isEmpty ? summary : "\(summary). \(detail)"
    }
}

File: android/app/src/main/kotlin/com/cabalmail/android/ui/feeds/FeedHealth.kt
package com.cabalmail.android.ui.feeds

import com.cabalmail.kit.models.RssFeedSummary

/** What the fetcher's view of a feed means for the sidebar badge (rss plan, phase 10, pulled forward). */
sealed interface FeedHealthLevel {
    data object Healthy : FeedHealthLevel

    data class Failing(
        val consecutiveFailures: Int,
    ) : FeedHealthLevel

    data object Stopped : FeedHealthLevel
}

/**
 * The health thresholds the Apple sidebar uses: a warning mark after three
 * consecutive fetch failures, a stop mark once the fetcher has given up
 * (twenty in a row, or dead-lettered). Pure, so the mapping is unit-tested
 * without Compose; the composables turn a level into words and a glyph.
 */
object FeedHealth {
    const val WARNING_THRESHOLD = 3
    const val STOPPED_THRESHOLD = 20

    fun level(feed: RssFeedSummary?): FeedHealthLevel {
        if (feed == null) return FeedHealthLevel.Healthy
        if (feed.deadLettered || feed.consecutiveFailureCount >= STOPPED_THRESHOLD) return FeedHealthLevel.Stopped
        if (feed.consecutiveFailureCount >= WARNING_THRESHOLD) {
            return FeedHealthLevel.Failing(feed.consecutiveFailureCount)
        }
        return FeedHealthLevel.Healthy
    }

    /**
     * The item list's header line: the level's [summary] followed by the
     * fetcher's own words when it left any. Null for a healthy feed.
     */
    fun headline(
        feed: RssFeedSummary?,
        summary: String?,
    ): String? {
        if (summary == null || feed == null) return null
        val words = feed.lastError.trim()
        return if (words.isEmpty()) summary else "$summary. $words"
    }
}
